### PR TITLE
Instrument tests: Avoid leaking state between tests

### PR DESCRIPTION
**Pull request by jenshnielsen**
When all pyvisa-sim instruments defined by a yaml file is closed reset the instruments. This in practice will ensure that every test will reset the state of pyvisa-sim instruments on exit since the test must close instruments anyway.

closes https://github.com/QCoDeS/Qcodes/issues/5356

See https://github.com/pyvisa/pyvisa-sim/issues/83 for details

### CODE REVIEW

File: qcodes/instrument/ip_to_visa.py [hunk 1]
```diff
@@ -90,6 +90,15 @@ def close(self) -> None:
...
             self.visa_handle.close()
 
+        if getattr(self, "visabackend", None) == "sim" and getattr(
+            self, "resource_manager", None
+        ):
+            # work around for https://github.com/pyvisa/pyvisa-sim/issues/83
+            # see other issues for more context
+            # https://github.com/QCoDeS/Qcodes/issues/5356 and
+            # https://github.com/pyvisa/pyvisa-sim/issues/82
+            self.resource_manager.visalib._init()
```
Comment (jenshnielsen) [inline qcodes/instrument/ip_to_visa.py:100]: Perhaps it would be safer if this is the last active session.

File: qcodes/instrument/ip_to_visa.py [hunk 2]
```diff
@@ -90,6 +90,23 @@ def close(self) -> None:
...
+            session_found = self.resource_manager.session in known_sessions
+
+            n_sessions = len(known_sessions)
+            # if this instrument is the last one or there are no connected instruments its safe to reset the device
+            if (session_found and n_sessions == 1) or n_sessions == 0:
+                # work around for https://github.com/pyvisa/pyvisa-sim/issues/83
+                # see other issues for more context
+                # https://github.com/QCoDeS/Qcodes/issues/5356 and
+                # https://github.com/pyvisa/pyvisa-sim/issues/82
+                self.resource_manager.visalib._init()
```
Comment (astafan8) [inline qcodes/instrument/ip_to_visa.py:108]: do we want to guard this with try/except for "_init" attribute defined in cacse pyvisa sim changes the implementation and/or name of this fucntion in the future? or is the damage very unlikely and only applicable to "sim" which is only used for tests and hence not in production, hence it's good as is?

**Review by astafan8 — APPROVED** [review]

File: qcodes/instrument/ip_to_visa.py (hunk 2 above)
Comment (jenshnielsen) [reply qcodes/instrument/ip_to_visa.py:108]: My original thought was that this is only used in testing but there is no harm in downgrading this to a warning if `_init` is not there.

Comment (astafan8) [reply qcodes/instrument/ip_to_visa.py:108]: > downgrading this to a warning if _init is not there.

i like this!

**Pull request merged**